mlflow/store/model_registry/file_store.py

```diff
@@ -0,0 +1,767 @@
+    ALL_STAGES,
+    DEFAULT_STAGES_FOR_GET_LATEST_VERSIONS,
+    STAGE_ARCHIVED,
+)
+from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_pb2 import (
+    INVALID_PARAMETER_VALUE,
+    RESOURCE_ALREADY_EXISTS,
+    RESOURCE_DOES_NOT_EXIST,
+)
+from mlflow.store.entities.paged_list import PagedList
+from mlflow.store.model_registry.abstract_store import AbstractStore
+from mlflow.store.model_registry import (
+    DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH,
+    SEARCH_REGISTERED_MODEL_MAX_RESULTS_THRESHOLD,
+)
+from mlflow.utils.search_utils import SearchUtils, SearchModelUtils, SearchModelVersionUtils
+from mlflow.utils.string_utils import is_string_type
+from mlflow.utils.validation import (
+    _validate_registered_model_tag,
+    _validate_model_version_tag,
+    _validate_model_name,
+    _validate_model_version,
+    _validate_tag_name,
+    _validate_param_name,
+)
+from mlflow.utils.env import get_env
+from mlflow.utils.file_utils import (
+    is_directory,
+    list_subdirs,
+    mkdir,
+    exists,
+    write_yaml,
+    read_yaml,
+    find,
+    read_file,
+    write_to,
+    make_containing_dirs,
+    list_all,
+    local_file_uri_to_path,
+)
+
+from .sqlalchemy_store import now
+
+_REGISTRY_DIR_ENV_VAR = "MLFLOW_REGISTRY_DIR"
+
+
+def _default_root_dir():
+    return get_env(_REGISTRY_DIR_ENV_VAR) or os.path.abspath(DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH)
+
+
+class FileStore(AbstractStore):
+    META_DATA_FILE_NAME = "meta.yaml"
+    TAGS_FOLDER_NAME = "tags"
+    MODEL_VERSION_TAGS_FOLDER_NAME = "tags"
+    CREATE_MODEL_VERSION_RETRIES = 3
+
+    def __init__(self, root_directory=None):
+        """
+        Create a new FileStore with the given root directory.
+        """
+
+        super().__init__()
+        self.root_directory = local_file_uri_to_path(root_directory or _default_root_dir())
+        # Create root directory if needed
+        if not exists(self.root_directory):
+            mkdir(self.root_directory)
+
+    def _check_root_dir(self):
+        """
+        Run checks before running directory operations.
+        """
+        if not exists(self.root_directory):
+            raise Exception("'%s' does not exist." % self.root_directory)
+        if not is_directory(self.root_directory):
+            raise Exception("'%s' is not a directory." % self.root_directory)
+
+    def _validate_registered_model_does_not_exist(self, name):
+        model = self.get_registered_model(name)
+        if model is not None:
+            raise MlflowException(
+                f"Registered Model (name={name}) already exists.",
+                RESOURCE_ALREADY_EXISTS,
+            )
+
+    def _update_registered_model_last_updated_time(self, name, updated_time):
+        registered_model = self._get_registered_model(name)
+        registered_model.last_updated_timestamp = updated_time
+        registered_model_dict = dict(registered_model)
+        del registered_model_dict["tags"]
+        write_yaml(
```

Comment:
We noticed in https://github.com/mlflow/mlflow/issues/6339 that, if a failure occurs during YAML write, file contents may be lost. As a result, we added https://github.com/mlflow/mlflow/blob/08205116b0844d43fdeef5062ff1d81304cce17a/mlflow/store/tracking/file_store.py#L998-L1026 to `FileStore` to write updates to temporary files and then use `shutil.move` to replace the preexisting file with the copied file. 

Can we turn that into a reusable utility, e.g. by placing it in `file_utils` and use it throughout this Model Registry FileStore?

Reply:
Sounds good!

Reply:
Saw this was resolved, but I don't think changes were pushed. Unresolving for now :)

Reply:
The implementation in https://github.com/mlflow/mlflow/pull/6605#discussion_r957813112 is not thread-safe. Shall we add file lock for it ? (don't need do it in this PR)